Readme.md
# Customized Jupyter scipy-notebook

Personalized version of https://hub.docker.com/r/jupyter/scipy-notebook/
with added:

- Jupyterlab extensions:  git, toc, nbdime
- ipympl:  for interactive plots
- interactive widgets

You can get the stock image, and install these yourself manually.
Note that leads to a larger image

## Why Docker?

Even more reproducible.
Instead of saying "get conda, install these packages, these extensions"
times N effort for each user,
plus documenting, possible environment conflicts,
tomorrow the instructions might not work etc.,
a docker image is immutable.
One person (or a few) can build and test, and it will be the same for everyone,
as long as you want.

Make local changes?  You can commit and name your version,
and if you break it (or the rest of the world breaks)
you can always get back to that and try something else.

Or run multiple copies, same or different versions.

## HOWTO use:

How depends on a few choices you make, depending on your workflow.

First you start the docker image, then you configure it / your environment.

### Start

Different cases:
1. run locally, just use localhost
1. server with ssl exposed

Orthogonal issues concerning work and data access
(hereafter just 'data"):
* keep notebooks in a local folder
* use a volume for data (shareable between containers?)
* user permissions for docker v your filesystem

#### <<example incantations to come>>
I run it like this (roughly):

    docker run -d -p $PORT:8888 \
       -e JUPYTER_ENABLE_LAB=yes \
       --name "my-scipy-plus" \
       -v "$(pwd)"/data:/data \
       -v "$(pwd)"/work:/home/jovyan/work  \
	 toddkaufmann/scipy-notebook-tk

This makes `data/` in current folder available as `/data` to docker / your notebooks,
and 'work/' will be the current working directory that jupyterlab is in on startup.

This way, all my work (data + notebooks & code) are outside the container.
I can remove or update the container without losing my work.
My environment (the python version, library, and modules) are locked in the container.
If I load/update modules here and break my environment somehow,
I can start again from the docker image.
I can run this image anywhere docker runs and not have to worry about the Python version, pip/conda installs, permissions, etc.
If I get attached to my environment, I can 'docker commit' to create a new image.

See `run.sh` as an example.

*Note*: `JUPYTER_ENABLE_LAB=yes` is the magic that turns on jupyterlab.

A docker tutorial is out of scope here, but will serve you well.

### Configure

For git:

Main thing for git is author info, if you want to commit.
Otherwise you will just get errors, I think they might only be shown as `error 128`.

Open a terminal in jupyter and run the appropriate `git config` commands, or copy your own `.gitconfig` to `/home/jovyan`.

### Troubleshooting

Q:  permission denied, etc.

A:  see if making world writable fixes it.


## Upgrading
1. got your data ?  If not in a separate folder/volume, make sure you copy that out.
2. maybe .gitconfig too; copy out
3. stop this container.
   (alternative:  run the new one on a diff port, you can leave this one running;
    useful for side-by-side comparison etc.)
4. start new, configure whatever you did before (.gitconfig, pip, etc)
   but encourage you to update Dockerfile with your changes as you make them
   (or make a request)
5. all good, you can remove the old if you want


## TODO:
- add example notebook (eg exploratory data analysis)
- include jq
- aws-cli, s3fs
- report bug (is it?) - when setting `NB_UID`
    COPY scipy-1226-tk2.2_start-notebook.sh  /usr/local/bin/start-notebook.sh 
- cleanup dockerfile, remove old comments (like this)
- animated notebook example


history
Thu Apr 25 20:33:36 EDT 2019 0.9.1 1 fff370d6d445
Thu Apr 25 20:44:05 EDT 2019 0.9.1 0 eda371791c7e
Sun Jun 16 19:24:40 EDT 2019 0.9.2 1 eda371791c7e
Sun Jun 16 19:44:49 EDT 2019 0.9.2 0 9b939a9cda64
Wed Sep 11 13:14:08 EDT 2019 0.9.3 0 0bbacce976fc
Fri Mar 6 13:10:33 EST 2020 2020.03.05 0 c9e786a8183a
Fri Mar 6 13:11:11 EST 2020 2020.03.05 1 c9e786a8183a
Fri Mar 6 13:24:17 EST 2020 2020.03.05 1 a206cebd758e
Wed May 13 10:17:26 EDT 2020 2020.05.13 0 a2833e0af29d
Sun Feb 7 20:21:35 EST 2021 2021-02-07.00 1 1257d543aed7
Sun Feb 7 20:23:30 EST 2021 2021-02-07.00 1 1257d543aed7
Sun Feb 7 20:38:01 EST 2021 2021-02-07.00 0 6b239233c1ec


run.sh
#!/bin/bash
set -eu
# docker run -p 8888:8888 jupyter/scipy-notebook
#IMAGE=tk/scipy-1128
# oops no tag/name
IMAGE=toddkaufmann/scipy-notebook-tk

# 'work' is actually where stuff is in the container.
# must be writable by uid 1000
# "Alternatively, specify the UID of the jovyan user on container
# startup using the -e NB_UID option described in the Common Features,
# Docker Options section"

# name you want to give your container
NAME=scipy-tk-0425

# local port to map to / expose:
PORT=8425
#JUPYTER_SERVER_URL="http://scipy.local:$PORT/"
JUPYTER_SERVER_URL="http://localhost:$PORT/"

# Make sure you are in the right place:
#  - I put work (notebooks) in 'work' folder
#  - data in 'data' folder
# and map them to local dirs outside of docker, for re-use, etc.
if ! [ -d work -a -d data ]; then
    echo not finding the directories
    exit 1
fi

echo ============================================
echo for NB_UID going to need to run as root supposedly
echo i.e.,  under --user root

# should be COPY'd by the dockerfile now
# echo also have to first start script ?
# docker cp scipy-1226-tk2.2_start-notebook.sh  ${IMAGE}:/usr/local/bin/start-notebook.sh 

#
# NOTE:
#  8888 is the port jupyter usually exposes
#  --user root:  works better with external volume for work/data
#  UID/GID:      reduce file permission issues; set to your own

docker run -d -p $PORT:8888 \
       --user root \
       -e NB_UID=1001 -e NB_GID=1001 \
       -e JUPYTER_ENABLE_LAB=yes \
       -e JUPYTER_SERVER_URL="$JUPYTER_SERVER_URL" \
       --name "$NAME" \
       -v "$(pwd)"/data:/data \
       -v "$(pwd)"/work:/home/jovyan/work  \
       $IMAGE

# Note:  above must have '-d' (detach)
# otherwise we won't get to this point (unless exit or detach (^P^Q ))

: 'first time' after run starts it ..
# we don't have to worry about .gitconfig already existing
if ! docker cp container.tmp/tk.gitconfig  ${NAME}:/home/jovyan/.gitconfig; then
    echo
    echo 'NOTE:'
    echo ' the git extension will fail until you configure it properly;'
    echo ' start a terminal in jupyter, cd to the appropriate folder and do "git init"'
    echo ' and appropriate "git config" commands to set your email/name.'
    echo '[by default, $HOME is inside the container and will lack the .gitconfig file]'
    echo
fi
#
# .bashrc - ? this is a good place for it



cat <<EOF
 .. watch the logs .. you will need the token to log in, 
 ..  then go to   $JUPYTER_SERVER_URL
 ..     ^C to exit any time after that.
=============================================================================
EOF

docker logs -f "$NAME"
exit 0
